Create Debian package for beszel-agent

Daniel15 opened the pull request:
Adds a Debian package to the build. Currently, only `beszel-agent` is packaged, but a package for `beszel-hub` could be added too. Since the SSH key is a required configuration option, the package prompts for it during installation.

This is the first step. The next step would be to set up a Debian repo for the packages.

# Test Plan
Run the build:
```bash
goreleaser release --snapshot --clean
```
This produces four `.deb` files in the `dist` directory, for amd64, arm64, armv6, riscv64, and mips64_hardfloat (architectures were already configured in the goreleaser config)

Create a new Debian VM
Install package:
```bash
cd /tmp
wget https://d.ls/debian/beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb
sudo apt install ./beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb
```

Output:
```
daniel@debtest:/tmp$ wget https://d.ls/debian/beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb
--2025-01-26 15:08:18--  https://d.ls/debian/beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb
Resolving d.ls (d.ls)... 185.197.30.201, 2a04:bdc7:100:ed8a::1
Connecting to d.ls (d.ls)|185.197.30.201|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3150376 (3.0M) [application/octet-stream]
Saving to: ‘beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb’

beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_am 100%[=============================================================================================>]   3.00M  --.-KB/s    in 0.1s    

2025-01-26 15:08:19 (23.5 MB/s) - ‘beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb’ saved [3150376/3150376]

daniel@debtest:/tmp$ sudo apt install ./beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'beszel-agent' instead of './beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb'
The following NEW packages will be installed:
  beszel-agent
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/3,150 kB of archives.
After this operation, 7,303 kB of additional disk space will be used.
Get:1 /tmp/beszel-agent_0.9.1-SNAPSHOT-e3aea63_linux_amd64.deb beszel-agent amd64 0.9.1~SNAPSHOT-e3aea63 [3,150 kB]
Preconfiguring packages ...
Selecting previously unselected package beszel-agent.
(Reading database ... 26393 files and directories currently installed.)
Preparing to unpack .../beszel-agent_0.9.1-SNAPSHOT-1747b00_linux_amd64.deb ...
Unpacking beszel-agent (0.9.1~SNAPSHOT-1747b00) ...
Setting up beszel-agent (0.9.1~SNAPSHOT-1747b00) ...
Creating beszel group
Creating beszel user
Created symlink /etc/systemd/system/multi-user.target.wants/beszel-agent.service → /lib/systemd/system/beszel-agent.service.
```

During the installation, it prompts for the SSH key:
![image](https://github.com/user-attachments/assets/e9280d56-09fa-4815-b471-9ce860e62aef)

Verify key is written to `/etc/beszel-agent.conf`:
```
daniel@debtest:/tmp$ sudo cat /etc/beszel-agent.conf 
KEY=ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIMiXhkOoriWaBgKbXa8/DTAR91nvRoi42AjRmTu1gsTG
```

Verify systemd unit is enabled and automatically started after installation:
```
daniel@debtest:/tmp$ sudo service beszel-agent status
● beszel-agent.service - Beszel Agent Service
     Loaded: loaded (/lib/systemd/system/beszel-agent.service; disabled; preset: enabled)
     Active: active (running) since Sun 2025-01-26 15:08:55 PST; 5min ago
   Main PID: 3982 (beszel-agent)
      Tasks: 6 (limit: 2315)
     Memory: 6.9M
        CPU: 79ms
     CGroup: /system.slice/beszel-agent.service
             └─3982 /usr/bin/beszel-agent

Jan 26 15:08:55 debtest systemd[1]: Started beszel-agent.service - Beszel Agent Service.
Jan 26 15:08:55 debtest beszel-agent[3982]: 2025/01/26 15:08:55 INFO Detected root device name=vda1
Jan 26 15:08:55 debtest beszel-agent[3982]: 2025/01/26 15:08:55 INFO Detected network interface name=enp1s0 sent=292698 recv=16869221
Jan 26 15:08:55 debtest beszel-agent[3982]: 2025/01/26 15:08:55 INFO Starting SSH server address=:45876
```

Verify connecting to it from Beszel Hub works:
![image](https://github.com/user-attachments/assets/2f31b33c-6850-409b-9ef7-07c56c514910)

Daniel15 commented:
Please don't merge yet as I still have to fix some Lintian warnings:
```
daniel@debtest:/tmp$ lintian beszel-agent_0.9.1-SNAPSHOT-e3aea63_linux_amd64.deb 
Warning in processable beszel-agent_0.9.1-SNAPSHOT-e3aea63_linux_amd64.deb: Use of uninitialized value $fraction in string eq at /usr/share/lintian/lib/Lintian/Check/Fields/Section.pm line 86.
Warning in processable beszel-agent_0.9.1-SNAPSHOT-e3aea63_linux_amd64.deb: Use of uninitialized value $fraction in string ne at /usr/share/lintian/lib/Lintian/Check/Fields/Section.pm line 95.
Warning in processable beszel-agent_0.9.1-SNAPSHOT-e3aea63_linux_amd64.deb: Use of uninitialized value $fraction in string eq at /usr/share/lintian/lib/Lintian/Check/Fields/Section.pm line 116.
E: beszel-agent: extended-description-is-empty
E: beszel-agent: maintainer-script-lacks-home-in-adduser "adduser --quiet --system "$SERVICE_USER" ??--ingroup "$SERVICE_USER" ??--no-create-home ??--gecos "System user for $SERVICE"" [postinst:30]
E: beszel-agent: malformed-contact Maintainer henrygd
E: beszel-agent: no-changelog usr/share/doc/beszel-agent/changelog.Debian.gz (non-native package)
E: beszel-agent: no-debconf-config
W: beszel-agent: empty-field Section
W: beszel-agent: hardening-no-pie [usr/bin/beszel-agent]
W: beszel-agent: hardening-no-relro [usr/bin/beszel-agent]
W: beszel-agent: maintainer-script-needs-depends-on-adduser adduser (does not satisfy adduser) [postinst:30]
W: beszel-agent: malformed-prompt-in-templates beszel-agent/key [templates:3]
W: beszel-agent: missing-debconf-dependency
W: beszel-agent: no-manual-page [usr/bin/beszel-agent]
W: beszel-agent: postinst-uses-db-input [postinst:15]
W: beszel-agent: postrm-does-not-purge-debconf postrm
W: beszel-agent: undeclared-elf-prerequisites (libc.so.6 libresolv.so.2) [usr/bin/beszel-agent]
W: beszel-agent: unknown-section

```

Daniel15 commented on supplemental/debian/beszel-agent.service:
```diff
@@ -0,0 +1,20 @@
+Restart=on-failure
+# TODO: See if we can apply more security hardening here
+# TODO: Audit using `systemd-analyze security beszel-agent.service`
+ProtectSystem=yes
```
These changes ( + whatever other harding options can be added) would be good to include in the install script too.

henrygd replied:
Great idea, I'll look over the other options.

Daniel15 replied:
@henrygd I ran `systemd-analyze security beszel-agent.service` but got overwhelmed at the number of different things it outputs. Maybe there's one flag we can set that sets all the standard security options. I'm not too familiar with these features of systemd.

Daniel15 commented on supplemental/debian/beszel-agent.service:
```diff
@@ -0,0 +1,20 @@
+
+[Service]
+Environment="PORT=45876"
+EnvironmentFile=/etc/beszel-agent.conf
```
Install script should probably do it this way too - store user-configured config environment variables in a separate file rather than directly in the systemd unit.

henrygd replied:
Agree, made a note to change this.

henrygd commented:
Beautiful, thank you!

Looks great, let me try it out tomorrow.

Daniel15 commented on supplemental/debian/lintian-overrides:
```diff
@@ -0,0 +1,11 @@
+beszel-agent: no-debconf-config
+beszel-agent: postinst-uses-db-input
+# Needs to be fixed in Beszel build
+beszel-agent: hardening-no-pie
```
From https://lintian.debian.org/tags/hardening-no-pie.html:

> This package provides an ELF executable that was not compiled as a position independent executable (PIE). PIE is required for fully enabling Address Space Layout Randomization (ASLR), which makes "Return-oriented" attacks more difficult.

I don't know Go, so I'm not sure how to do this.

henrygd replied:
Never ran into this before but it's probably some build flag that it needs. I'll look into it.

henrygd commented on supplemental/debian/lintian-overrides:
```diff
@@ -0,0 +1,11 @@
+# Needs to be fixed in Beszel build
+beszel-agent: hardening-no-pie
+beszel-agent: hardening-no-relro
+# Maybe one day
```
:laughing:

henrygd commented:
Ran out of time to look at this tonight but I'll try to finish up tomorrow if possible.

Worked nicely in an Ubuntu VM. I did notice that neither `apt remove` or `apt purge` actually stopped the running service. I assume we need to do that in the uninstall script?

After going through the security options I think these should be fine to set for the time being. And then probably add `RuntimeDirectory=beszel/beszel-agent` to leave us the potential to write to `/run/beszel/` in the future if we add a sqlite database or something. 

```
KeyringMode=private
LockPersonality=yes
NoNewPrivileges=yes
PrivateTmp=yes
ProtectClock=yes
ProtectHome=read-only
ProtectHostname=yes
ProtectKernel=yes
ProtectKernelLogs=yes
ProtectKernelTunables=yes
ProtectSystem=strict
ReadWritePaths=/run
RemoveIPC=yes
RestrictSUIDSGID=true
SystemCallArchitectures=native
```

I think the `hardening-no-pie` thing is specific to the riscv build, as riscv apparently requires external linking with Cgo to use `-buildmode=pie`. The only way around that is to compile directly on riscv, which isn't possible. 

I'd like to just exclude riscv and mips from the deb builds, but there doesn't seem to be a way to do that without a monthly subscription to GoReleaser Pro. :shrug:

Daniel15 commented:
Thanks for looking into it, and thanks for testing on Ubuntu since I only tested on Debian. I think you're right that the service needs to be stopped on removal. I'll add that soon (by end of week). 

A SQLite database would go in /var/lib. /run is for things that don't need to be persisted and are okay to delete after rebooting (PID files, sockets, etc)



On January 27, 2025 7:09:34 PM PST, hank ***@***.***> wrote:
>Ran out of time to look at this tonight but I'll try to finish up tomorrow if possible.
>
>Worked nicely in an Ubuntu VM. I did notice that neither `apt remove` or `apt purge` actually stopped the running service. I assume we need to do that in the uninstall script?
>
>After going through the security options I think these should be fine to set for the time being. And then probably add `RuntimeDirectory=beszel/beszel-agent` to leave us the potential to write to `/run/beszel/` in the future if we add a sqlite database or something. 
>
>```
>KeyringMode=private
>LockPersonality=yes
>NoNewPrivileges=yes
>PrivateTmp=yes
>ProtectClock=yes
>ProtectHome=read-only
>ProtectHostname=yes
>ProtectKernel=yes
>ProtectKernelLogs=yes
>ProtectKernelTunables=yes
>ProtectSystem=strict
>ReadWritePaths=/run
>RemoveIPC=yes
>RestrictSUIDSGID=true
>SystemCallArchitectures=native
>```
>
>I think the `hardening-no-pie` thing is specific to the riscv build, as riscv apparently requires external linking with Cgo to use `-buildmode=pie`. The only way around that is to compile directly on riscv, which isn't possible. 
>
>I'd like to just exclude riscv and mips from the deb builds, but there doesn't seem to be a way to do that without a monthly subscription to GoReleaser Pro. :shrug:
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/henrygd/beszel/pull/497#issuecomment-2617746496
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>

henrygd commented:
True, looks like `StateDirectory` is the best option as it creates a directory in `/var/lib`.

```ini
KeyringMode=private
LockPersonality=yes
NoNewPrivileges=yes
PrivateTmp=yes
ProtectClock=yes
ProtectHome=read-only
ProtectHostname=yes
ProtectKernel=yes
ProtectKernelLogs=yes
ProtectKernelTunables=yes
ProtectSystem=strict
RemoveIPC=yes
RestrictSUIDSGID=true
StateDirectory=beszel-agent
SystemCallArchitectures=native
```

Daniel15 commented:
I've implemented the requested changes. It still seems to be working fine after the changes to the systemd unit. I also added a pre-removal script that stops the systemd service, and verified that it stops properly, and that subsequent reinstallation works as expected.

henrygd commented:
Merged, thanks!

I also updated the docs and install script with the new security options.

Adding a changelog is on my todo list so I'll get to it before too long.

Daniel15 commented:
Thanks!

Justinzobel commented:
Is there an issue/PR for tracking packaging the Hub? I am wanting to install this into production but would rather a Debian package for easy/clean install/upgrade/removal.

Daniel15 commented:
Not at the moment. I didn't have time to implement both, and chose to focus on the agent since you install the agent on far more systems than the hub.

You'd essentially need to do the same thing as this PR. Some of it can be reused. The scripts in `supplemental` would need to be renamed (e.g. `supplemental/debian/prerm.sh` to `supplemental/debian/prerm-agent.sh`) or moved to a subdirectory (e.g. `supplemental/debian/agent/prerm.sh`) since their contents will differ between the two packages.
